### PR TITLE
Make ZeppelinContext extensible like in version 0.8.0-SNAPSHOT and available in Flink Interpreter

### PR DESCRIPTION
**Pull request by marcotagliabue**
### What is this PR for?
ZeppelinContext has only supported by Spark Interpreter. This PR  makes it extensible to other interpreters and in particular it's been implemented for Flink.

### What type of PR is it?
[Feature]

### What is the Jira issue?

https://issues.apache.org/jira/browse/ZEPPELIN-1596
https://issues.apache.org/jira/browse/ZEPPELIN-1595

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

File: flink/src/main/java/org/apache/zeppelin/flink/FlinkZeppelinContext.java [hunk 1]
```diff
@@ -0,0 +1,810 @@
+package org.apache.zeppelin.flink;
+
```
Comment (maocorte) [inline flink/src/main/java/org/apache/zeppelin/flink/FlinkZeppelinContext.java:19]: missing apache license header

File: flink/src/main/java/org/apache/zeppelin/flink/FlinkInterpreter.java [hunk 2]
```diff
@@ -60,17 +55,25 @@
...
   private LocalFlinkMiniCluster localFlinkCluster;
+  private IMain imain;
   private FlinkILoop flinkIloop;
+  private FlinkZeppelinContext zFlink;
+  private ByteArrayOutputStream out;
+
   private Map<String, Object> binder;
-  private IMain imain;
+  private static InterpreterHookRegistry hooks;
+
```
Comment (robbenti) [inline flink/src/main/java/org/apache/zeppelin/flink/FlinkInterpreter.java]: Remove this empty line

File: flink/src/main/java/org/apache/zeppelin/flink/FlinkInterpreter.java [hunk 3]
```diff
@@ -193,6 +208,7 @@ private Settings createSettings() {
     settings.scala$tools$nsc$settings$ScalaSettings$_setter_$maxClassfileName_$eq(
             numClassFileSetting);
 
+
```
Comment (robbenti) [inline flink/src/main/java/org/apache/zeppelin/flink/FlinkInterpreter.java]: Remove this empty line

File: flink/src/main/java/org/apache/zeppelin/flink/FlinkZeppelinContext.java [hunk 4]
```diff
@@ -0,0 +1,810 @@
...
+    interpreterClassMap = new HashMap<String, String>();
+    interpreterClassMap.put("flink", "org.apache.zeppelin.flink.FlinkInterpreter");
+
+    this.supportedClasses = new ArrayList<>();
+    try {
+      supportedClasses.add(
+          this.getClass().forName("org.apache.flink.streaming.api.scala.DataStream"));
+    } catch (ClassNotFoundException e) {
+    }
```
Comment (robbenti) [inline flink/src/main/java/org/apache/zeppelin/flink/FlinkZeppelinContext.java]: At least log the error

File: flink/src/main/java/org/apache/zeppelin/flink/FlinkZeppelinContext.java [hunk 5]
```diff
@@ -0,0 +1,810 @@
...
+
+    try {
+      supportedClasses.add(
+          this.getClass().forName("org.apache.flink.streaming.api.scala.DataSet"));
+    } catch (ClassNotFoundException e) {
+    }
```
Comment (robbenti) [inline flink/src/main/java/org/apache/zeppelin/flink/FlinkZeppelinContext.java]: At least log the error

**Review by maocorte — APPROVED** [review]

**Review by robbenti — APPROVED** [review]
LGTM

**Pull request merged**